mxnet-bot/LabelBotAddLabels/label_bot.py

```diff
@@ -0,0 +1,143 @@
+def tokenize(string):
+    substring = string[string.find('[')+1: string.rfind(']')] 
+    labels = [' '.join(label.split()) for label in substring.split(',')]
+    logger.info("recognize labels: {}".format(", ".join(labels)))
+    return labels
+
+
+# covert all non-alphanumeric characters from raw_string to sub_string
+def clean_string(raw_string, sub_string):
+    cleans = re.sub("[^0-9a-zA-Z]", sub_string, raw_string)
+    return cleans.lower()
+
+# This method is to count how many pages of issues/labels in total
+def count_pages(obj, state='all'):
+    # obj could be "issues"/"labels"
+    # state could be "open"/"closed"/"all", available to issues
+    assert obj in set(["issues", "labels"]), "Invalid Input!"
+    url = 'https://api.github.com/repos/{}/{}'.format(REPO, obj)
+    if obj == 'issues':
+        response = requests.get(url, {'state': state},
+                                auth=AUTH)
+    else:
+        response = requests.get(url, auth=AUTH)
+    assert response.status_code == 200, response.status_code
+    if "link" not in response.headers:
+        return 1
+    # response.headers['link'] will looks like:
+    # <https://api.github.com/repositories/34864402/issues?state=all&page=387>; rel="last"
+    # In this case we need to extrac '387' as the count of pages
+    return int(clean_string(response.headers['link'], " ").split()[-3])
+
+
+# This method is to find comments which @mxnet-label-bot
+def find_notifications():
+    data = []
+    pages = count_pages("issues")
+    for page in range(1, pages+1):
+        url = 'https://api.github.com/repos/' + REPO + '/issues?page=' + str(page) \
+            + '&per_page=30'.format(repo=REPO)
+        response = requests.get(url,
+                                {'state': 'open',
+                                 'base': 'master',
+                                 'sort': 'created',
+                                 'direction': 'desc'},
+                                auth=AUTH)
+        for item in response.json():
+            # limit the amount of unlabeled issues per execution
+            if len(data) >= 50:
+                break
+            if "pull_request" in item:
+                continue
+            if not item['labels']:
+                if item['comments'] != 0:
+                    labels = []
+                    comments_url = "https://api.github.com/repos/{}/issues/{}/comments".format(REPO,item['number'])
+                    comments = requests.get(comments_url, auth=AUTH).json()
+                    for comment in comments:
+                        if "@mxnet-label-bot" in comment['body']:
+                            labels += tokenize(comment['body'])
+                            logger.info("issue: {}, comment: {}".format(str(item['number']),comment['body']))
+                    if labels != []:
+                        data.append({"issue": item['number'], "labels": labels})
+    return data
+
+
+# find all existing labels in a repo
+def all_labels():
```

Comment:
can we pass repo / AUTH as a parameter? makes the function testable.

Reply:
We can either make it a class where repo auth are taken from constructor or have the functions take parameters. This using globals is not very good style and is not testable.

Reply:
Refactored this file into a class